### PR TITLE
add types for ShExJ

### PR DESCRIPTION
**Pull request by ericprud**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

See JSG specification for ShExJ at http://shex.io/shex-semantics/#shexj .

### CODE REVIEW

File: types/shexj/tslint.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "extends": "dtslint/dt.json",
+    "rules": {
```
Comment (sheetalkamat) [inline types/shexj/tslint.json:3]: You cannot disable this rule. Please remove this and fix the error

File: types/shexj/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,576 @@
+// Type definitions for ShExJ 2.1
```
Comment (sheetalkamat) [inline types/shexj/index.d.ts:1]: npmjs does not show any package with this name. so this needs to be non npm package

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/shexj/index.d.ts (hunk 2 above)
Comment (ericprud) [reply types/shexj/index.d.ts:1]: There's a constellation of ShEx packages in the `@shexjs` scope. They provide code to manipulate ShExJ structures. These structures were [formally modelled](http://shex.io/shex-semantics/#shexj) but never before in Typescript. There are some other packages which use ShEx as well. The idea is that `@types/shexj` be a standard structure all (ts-aware) ShExJ tools can share.

I think the closest analogy would be `@types/rdf-js` which defines structures and interfaces used by e.g. `n3`, `graphy` etc. Likewise, `@shexjs/core`, `@shexjs/parser`, and a lot of pending code in [solid](https://github.com/solid/solid) will leverage this typing.

Given that, is it necessary to overlap with an existing npm name?

Comment (rubensworks) [reply types/shexj/index.d.ts:1]: For the `rdf-js` typings, it was [suggested](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/20507#issuecomment-336071454) that the npm package would also be claimed (which I did). Just to make sure that the TS tooling would not break. So I think it's a good idea to claim `shexjs` just like [`rdf-js`](https://www.npmjs.com/package/rdf-js).

**Review by rubensworks — APPROVED** [review]

File: types/shexj/index.d.ts (hunk 2 above)
Comment (ericprud) [reply types/shexj/index.d.ts:1]: I think I've take ownership two lines below per [Definition owners](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#definition-owners).
Is some other speech act required?

Comment (sheetalkamat) [reply types/shexj/index.d.ts:1]: If you remove the rule linter will show you how to fix that error by updating header comment

Comment (ericprud) [reply types/shexj/index.d.ts]: gotcha! ok, fixed.

The linter was somewhat helpful in that it suggested adding:

> // Type definitions for non-npm package shexj-browser

In reality, I had to replace the existing header line with a header with "non-npm package" and a version:

> // Type definitions for non-npm package ShExJ 2.1

File: types/shexj/tslint.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+        "npm-naming": true
+    }
```
Comment (sheetalkamat) [inline types/shexj/tslint.json:5]:
```suggestion
```

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

**Review by rubensworks — APPROVED** [review]

**Comment by ericprud** [comment]
I think this should be good to go. I accepted @sheetalkamat 's edits and made it compile.

**Review by peterblazejewicz — APPROVED** [review]

**Comment by ericprud** [comment]
The issue summary says there's still 1 change requested. As far as I can tell, it was to remove ` "rules": {}` from tslint.json, which I believe to be [done](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50899/files#diff-5769d54b1994f59a2e4e18d3b3d2f8114cb718d2210b5fec1ae4646d9e07112e).

**Review by peterblazejewicz — APPROVED** [review]

**Review by rbuckton — APPROVED** [review]

**Comment by ericprud** [comment]
Ready to merge

**Pull request merged**